test/lib-test/jdk/test/lib/hprof/HprofTest.java

```diff
@@ -0,0 +1,171 @@
+                }
+            }
+
+            if (p.exitValue() != 0) {
+                throw new Exception("Jcmd exited with code " + p.exitValue());
+            }
+        } finally {
+            LingeredApp.stopApp(theApp);
+        }
+    }
+
+    private static void test(File dumpFile) throws Exception {
+        Asserts.assertTrue(dumpFile.exists(), "Heap dump file not found.");
+
+        log("Reading " + dumpFile + "...");
+        try (Snapshot snapshot = Reader.readFile(dumpFile.getPath(), true, 0)) {
+            log("Resolving snapshot...");
+            snapshot.resolve(true);
+            log("Snapshot resolved.");
+
+            JavaObject testObj = getTestObject(snapshot);
+            testPrimitiveArray(testObj);
+            testString(testObj);
+        }
+
+    }
+
+    // verifies JavaValueArray.valueString does not throw
+    // "invalid array element type" exception
+    private static void testPrimitiveArray(JavaObject obj) {
+        JavaHeapObject field = getObjectField(obj, "intArray");
+        Asserts.assertTrue(field instanceof JavaValueArray);
+        log("int array: " + ((JavaValueArray)field).valueString());
+    }
+
+    // verifies JavaObject.toString returns String value
+    private static void testString(JavaObject obj) {
+        JavaHeapObject field = getObjectField(obj, "str");
+        Asserts.assertTrue(field instanceof JavaObject);
+        JavaObject javaObj = (JavaObject)field;
+        Asserts.assertTrue(javaObj.getClazz().isString());
+        log("string: " + javaObj.toString());
+        assert(javaObj.toString().contains(new HprofTestTarg().str));
+    }
+
+
+    private static JavaHeapObject getObjectField(JavaObject obj, String fieldName) {
+        JavaThing thing = obj.getField(fieldName);
+        // only non-primitive types are supported
+        return (JavaHeapObject)thing;
+    }
+
+    // gets test HprofTestTarg
+    private static JavaObject getTestObject(Snapshot snapshot) {
+        String testClassName = HprofTestTarg.class.getName();
+        JavaHeapObject testObject = getObjects(snapshot, testClassName).nextElement();
+        Asserts.assertTrue(testObject instanceof JavaObject);
+        return (JavaObject)testObject;
+    }
+
+    // finds all objects of the specified type
+    private static Enumeration<JavaHeapObject> getObjects(Snapshot snapshot, String className) {
+        log("Looking for '" + className + "' objects...");
+        JavaClass jClass = snapshot.findClass(className);
+        if (jClass == null) {
+            throw new RuntimeException("Class '" + className + "' not found");
+        }
+        int instanceCount = jClass.getInstancesCount(false);
+        if (instanceCount < 1) {
+            throw new RuntimeException("Not instances of '" + className + "' found");
+        }
+        log("Found " + instanceCount + " instance(s).");
+        return jClass.getInstances(false);
+    }
+
+    private static void log(Object s) {
```

Comment:
Nit: Why is the arg of `Object` type? Also, all calls to `log()` always pass `String` type.

Reply:
Sometimes it it's useful to have ability to log arbitrary objects (or primitive values which are auto-boxed) without explicit conversion to String